### DISCUSSION TITLE
How to get the direction-dependent material properties?

### DISCUSSION THREAD

**Discussion by rh201** [idaholab/moose#20312] [Q&A General] [answered]
Hi MOOSEers,

When I try to use direction-dependent material properties like the form below, how can I calculate the angle of the normal vector w.r.t the x/y-axis and get direction (\phi) here ? Is there anything related on it? I appreciate any suggestions and help.Thank you.

![image](https://user-images.githubusercontent.com/79764031/154156237-b69f0479-c213-40b9-8eec-675f1af4ce45.png)

**Comment by GiudGiud** [comment, marked as answer]
Hello

When you solve the equation do you always evaluate the property on the X/Y/Z axis or can it be any direction?
We usually represent anisotropy with a tensor material property

Guillaume

**Reply by rh201** [reply]
Hi~ 

If I want to get \phi here, I need to compute the projection on X/Y/Z, am I right?

Could you refer me to the tensor material property you mentioned? I'm interested in it and want to see how it works. Thank you.

**Reply by GiudGiud** [reply]
Here's an example of a rank two (matrix) tensor property
https://mooseframework.inl.gov/source/materials/ADGenericFunctionRankTwoTensor.html

If there is no cross terms you can also just use a vector. 

So to compute phi in your case, it seems D (is it a variable?) 's gradient should be passed to the material property (using coupledGradient).
Then e2 and e1 are fixed I imagine, so computing phi should be ok.

G is going to be a material property? Or is the equation you pasted the equation for the non linear system you're trying to solve?

**Reply by rh201** [reply]
Thank you. I'll look into that page. @GiudGiud 

Yes, e1 and e2 is the unit vector of X/Y axis, D is the variable and Gc is the material property which I want it to be \phi-dependent. 
I tried to calculate the D's gradient and use Gc directly in the kernel (with AD) as follows:

>     ADRealVectorValue x_unit = RealVectorValue(1,0,0);
>     ADRealVectorValue y_unit = RealVectorValue(0,1,0);
>     ADReal x_Projection =   _grad_u[_qp] * x_unit;
>     ADReal y_Projection =   _grad_u[_qp] * y_unit;
>     ADReal theta = std::atan2(y_Projection, x_Projection) - pi/2 ;
>     return _Gc[_qp] * std::sin(theta) * ...

It compiles but I'm not sure of it. Could you help me check that? Thank you.

**Reply by GiudGiud** [reply]
Looks fine to me.
You can initialize x_unit and y_unit like this:
`` RealVectorValue x_unit(1.0, 0.0, 0.0); ``

**Reply by rh201** [reply]
Thank you. I'll give it a try.

Bests.
